### ISSUE TITLE
Recovering old data from a GCP bucket

### ISSUE THREAD

**Issue by nostdm** [upspin/upspin#597]
**What version of Upspin are you using?**

Build time: May 17 06:03:43 2018 UTC
Git hash:   63f1073c7a3a0fe6162849630916771c15ed80e9

**What operating system and processor architecture are you using?**

Ubuntu 18.04 AMD64

**Issue description**

I set up Upspin about a year ago, recently my GCP promotional period ended and Google decided it was a good idea to delete my Compute instances, including my `upspinserver` instance. Now all I have left is the old GCP Bucket with upspin storage objects.

Due to the hassle associated with setting up the server from scratch, I instead used the `upspin-ui` command to set up a new upspinserver on GCP. I then SSH'ed into the new server instance and modified the `serverconfig.json` to point to the old bucket and changed the username to be the same as the old username. I also updated my local config to reflect the old username.

Connection-wise everything seems to work as expected, but after restarting the server, systemctl reports:

```
● upspinserver.service - An upspinserver container instance
   Loaded: loaded (/etc/systemd/system/upspinserver.service; static; vendor preset: disabled)
   Active: active (running) since Sat 2018-06-30 00:23:17 UTC; 21min ago
  Process: 646 ExecStartPre=/usr/bin/docker pull gcr.io/upspin-containers/upspinserver:latest (code=exited, status=0/SUCCESS)
  Process: 638 ExecStartPre=/usr/bin/docker-credential-gcr configure-docker (code=exited, status=0/SUCCESS)
 Main PID: 659 (docker)
    Tasks: 5 (limit: 4915)
   Memory: 23.1M
      CPU: 175ms
   CGroup: /system.slice/upspinserver.service
           └─659 /usr/bin/docker run --rm -u=2000 --volume=/home/upspin:/upspin -p=80:8080 -p=443:8443 --name=upspinserver gcr.io/upspin-containers/upspinserver:latest
Jun 30 00:43:48 upspinserver docker[659]:         user example.user+upspinserver@gmail.com:
Jun 30 00:43:48 upspinserver docker[659]:         no root for user
Jun 30 00:44:18 upspinserver docker[659]: 2018/06/30 00:44:18.836167 serverutil/perm.NewFromDir: item does not exist:
Jun 30 00:44:18 upspinserver docker[659]:         dir/server.Watch:
Jun 30 00:44:18 upspinserver docker[659]:         user example.user+upspinserver@gmail.com:
Jun 30 00:44:18 upspinserver docker[659]:         no root for user
```
The old bucket seems to contain a root file:
```
tree.root.example.user@gmail.com.03465d7842682442
```
But yet when I do `upspin ls example.user@gmail.com` it returns the following:

```
$ upspin ls example.user@gmail.com
upspin: ls: dir/remote("9adbea87ccebc4cf404cf121868646c3.upspin.services:443").Lookup: item does not exist:
	dir/server.Lookup:
	user example.user@gmail.com:
	no root for user
```

I seem to be able to create new files with `echo Hello, Upspin | upspin put example.user@gmail.com/hello` and then read them `upspin get example.user@gmail.com/hello`

But the old files are nowhere to be found. Am I doing something wrong?

Also, is there a way of manually recovering the upspin data directly from the Bucket objects?

The other bucket that I created later, contains two root files, but the original one contains only a single one, could that be an issue:
```
tree.root.example2.user@gmail.com.03465d7842682442
tree.root.example2.user+upspinserver@gmail.com.03465d7842682442
```
Thanks a lot.

**Comment by nostdm** [comment]
Sorry for the late response, I still haven't managed to solve this issue. Today I spent most of the day trying to set up `upspinserver` from scratch and I managed to set up a dockerized (`gcr.io/upspin-containers/upspinserver`) upspin server with a fresh GCP bucket.

That said, when trying to re-use my old bucket I'm having trouble making the server understand how to read the old data (which kind of makes me question the decision to use upspin to store all of my data if it's this difficult to retrieve it once the server configuration is deleted). Regardless, this is what I'm currently doing which isn't working:

1. I run `upspin setupdomain --domain=mydomain.com` which if I understood correctly sets up the server user to be `upspin@mydomain.com` instead of the default `my.user.email+upspinserver@gmail.com` and creates a local `deploy` directory with server configuration.

2. Then I run `upspin setupstorage-gcp -domain=mydomain.com --project=myproject-xyz my-old-bucket` the output of the command is:
```
$ upspin setupstorage-gcp -domain=mydomain.com --project=myproject-xyz my-old-bucket
A new key for the service account "upspinstorage@myproject-xyz.iam.gserviceaccount.com" was created.
Bucket "my-old-bucket" already exists; re-using it.
You should now deploy the upspinserver binary and run 'upspin setupserver'.
```
If I understood correctly this step creates a corresponding service account to access the bucket objects and adds the configuration to the `serverconfig.json` file in the `deploy` directory which will later be pushed to the server.

3. I run `upspin setupserver --domain=mydomain.com --host=upspin.mydomain.com` where the host has been previously registered with the upspin keyserver with a DNS record `upspin.mydomain.com` that points to the server's IP address. The output of the command is:
```
Successfully put "upspin@mydomain.com" to the key server.
Configured upspinserver at "upspin.mydomain.com:443".
```

If I understood correctly this step pushes the generated configuration files to the server and points the server at the correct GCP bucket.

However, after running the last command, the first issue seems to be service account permissions, despite having owner access to the bucket, I still get permission denied errors:
```
upspin: setupserver: client.MakeDirectory: my.user.email@gmail.com/:
	dir/remote("upspin.mydomain.com:443").Put:
	dir/server.Put:
	cloud/storage/gcs.Download: googleapi: got HTTP response code 403 with body: upspinstorage@myproject-xyz.iam.gserviceaccount.com does not have storage.objects.get access to my-old-bucket/tree.root.my.user.email@gmail.com.bb24f6bc4737ed51.
```
Second issue is that the `tree.root` file that is logged in the error has a different ending than the old `tree.root` file despite being the same user account.

Can anyone tell me what I'm doing wrong here? How do I make the server use the old `tree.root` where it seems to be wanting to set up a new one?

**Comment by nostdm** [comment]
Further update: 

I managed to fix the permission issues by manually adding the `upspinstorage@myproject-xyz.iam.gserviceaccount.com` to the bucket permissions and setting the permission level to owner.

After running the server it creates two new root files in the bucket, one for the user account and the other for the server account.

In order to re-use the old root file, I copied the `tree.root.my.user.email@gmail.com` file to the server's `dirserver-logs` directory by replacing the already existing file with the same name `tree.root.my.user.email@gmail.com` (I left the index files as they were). 

But after restarting the server and running `echo "Hello world" | upspin put daniel.chmiel.dev@gmail.com/hello` the output I get is:
```
upspin: put: dir/remote("upspin.mydomain.com:443").Put: no wrapped key for user; owner must "upspin share -fix":
	dir/server.Put: my.user.email@gmail.com/hello:
	my.user.email@gmail.com/:
	pack/ee.Unpack: user upspin@mydomain.com
```
